File: IOTConnectSDK/iotconnect-sdk/src/main/java/com/iotconnectsdk/mqtt/IotSDKMQTTService.java
package com.iotconnectsdk.mqtt;

import android.content.Context;

import com.iotconnectsdk.R;
import com.iotconnectsdk.interfaces.HubToSdkCallback;
import com.iotconnectsdk.interfaces.TwinUpdateCallback;
import com.iotconnectsdk.utils.IotSDKLogUtils;
import com.iotconnectsdk.utils.IotSDKUtils;
import com.iotconnectsdk.webservices.responsebean.SyncServiceResponse;

import org.eclipse.paho.android.service.MqttAndroidClient;
import org.eclipse.paho.client.mqttv3.DisconnectedBufferOptions;
import org.eclipse.paho.client.mqttv3.IMqttActionListener;
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.IMqttToken;
import org.eclipse.paho.client.mqttv3.MqttCallbackExtended;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;
import org.json.JSONObject;

/**
 * Service class for MQTT data upload to cloud
 */
public class IotSDKMQTTService {

    private static final String TAG = IotSDKMQTTService.class.getSimpleName();

    private MqttAndroidClient mqttAndroidClient;
    private final HubToSdkCallback mHubToSdkCallback;
    private final TwinUpdateCallback twinCallbackMessage;

    //- To receive the desired property and update the Reported Property
    private final String TWIN_PUB_TOPIC = "$iothub/twin/PATCH/properties/reported/?$rid=1";
    private final String TWIN_SUB_TOPIC = "$iothub/twin/PATCH/properties/desired/#";


    // - To Publish the blank message on publish and Subscribe message with desired and reported property
    private final String TWIN_PUB_TOPIC_BLANK_MSG = "$iothub/twin/GET/?$rid=0";
    private final String TWIN_SUB_TOPIC_BLANK_MSG = "$iothub/twin/res/#";

    private final String DESIRED = "desired";
    private final String UNIQUE_ID = "uniqueId";

    private final String TWIN_SUBTOPIC_CONTAINT = "$iothub/twin/";

    private String subscriptionTopic;// = "devices/520uta-sdk003/messages/devicebound/#";
    private String publishTopic;// = "devices/520uta-sdk003/messages/events/";

    private final IotSDKLogUtils iotSDKLogUtils;
    private final boolean isDebug;
    private final Context context;
    private final String uniqueId;
    private final SyncServiceResponse.DBeanXX.PBean protocolBean;


    private static IotSDKMQTTService iotSDKMQTTService;

    public static IotSDKMQTTService getInstance(Context context, SyncServiceResponse.DBeanXX.PBean protocolBean, HubToSdkCallback hubToSdkCallback, final TwinUpdateCallback twinCallbackMessage, IotSDKLogUtils iotSDKLogUtils, boolean isDebug, String uniqueId) {

        if (iotSDKMQTTService == null) {
            return iotSDKMQTTService = new IotSDKMQTTService(context, protocolBean, hubToSdkCallback, twinCallbackMessage, iotSDKLogUtils, isDebug, uniqueId);
        }
        return iotSDKMQTTService;
    }

    public void clearInstance() {
        iotSDKMQTTService = null;
    }

    private IotSDKMQTTService(Context context, SyncServiceResponse.DBeanXX.PBean protocolBean, HubToSdkCallback hubToSdkCallback, final TwinUpdateCallback twinCallbackMessage, IotSDKLogUtils iotSDKLogUtils, boolean isDebug, String uniqueId) {
        this.context = context;
        this.uniqueId = uniqueId;
        this.protocolBean = protocolBean;
        this.mHubToSdkCallback = hubToSdkCallback;
        this.twinCallbackMessage = twinCallbackMessage;
        this.iotSDKLogUtils = iotSDKLogUtils;
        this.isDebug = isDebug;
    }

    public void connectMQTT() {
        //init log.
        iotSDKLogUtils.log(false, this.isDebug, "INFO_IN04", context.getString(R.string.INFO_IN04));

        subscriptionTopic = protocolBean.getSub();
        publishTopic = protocolBean.getPub();

        mqttAndroidClient = new MqttAndroidClient(context, "ssl://" + this.protocolBean.getH() + ":" + this.protocolBean.getP(), this.protocolBean.getId());

        mqttAndroidClient.setCallback(new MqttCallbackExtended() {
            @Override
            public void connectComplete(boolean reconnect, String serverURI) {
                //if (reconnect) {
                    // Because Clean Session is true, we need to re-subscribe
                    if (subscriptionTopic != null) {
                        subscribeToTopic();
                    }
                /*} else {
//                    addToHistory("Connected to: " + serverURI);
                }*/
            }

            @Override
            public void connectionLost(Throwable cause) {
                mHubToSdkCallback.onConnectionStateChange(false);
            }

            @Override
            public void messageArrived(String topic, MqttMessage message) throws Exception {

                try {
                    if (message.getPayload().toString().isEmpty()) {
                        return;
                    }

                    if (topic.contains(TWIN_SUB_TOPIC.substring(0, TWIN_SUB_TOPIC.length() - 1))) {
                        JSONObject mainObj = new JSONObject();
                        mainObj.put(DESIRED, new JSONObject(new String(message.getPayload())));
                        mainObj.put(UNIQUE_ID, uniqueId);

                        twinCallbackMessage.twinUpdateCallback(mainObj);

                    } else if (topic.contains(TWIN_SUB_TOPIC_BLANK_MSG.substring(0, TWIN_SUB_TOPIC_BLANK_MSG.length() - 1))) {

                        JSONObject mainObj = new JSONObject(new String(message.getPayload()));
                        mainObj.put(UNIQUE_ID, uniqueId);

                        twinCallbackMessage.twinUpdateCallback(mainObj);
                    } else {
                        mHubToSdkCallback.onReceiveMsg(new String(message.getPayload()));
                    }

                } catch (Exception e) {
                    e.printStackTrace();
                }

            }

            @Override
            public void deliveryComplete(IMqttDeliveryToken token) {
            }
        });

        MqttConnectOptions mqttConnectOptions = new MqttConnectOptions();
        mqttConnectOptions.setAutomaticReconnect(true);
        mqttConnectOptions.setCleanSession(true);
        mqttConnectOptions.setUserName(protocolBean.getUn());
        if (protocolBean.getPwd() != null) {
            mqttConnectOptions.setPassword(protocolBean.getPwd().toCharArray());
        }

        try {
            mqttAndroidClient.connect(mqttConnectOptions, null, new IMqttActionListener() {

                @Override
                public void onSuccess(IMqttToken asyncActionToken) {
                    DisconnectedBufferOptions disconnectedBufferOptions = new DisconnectedBufferOptions();
                    disconnectedBufferOptions.setBufferEnabled(true);
                    disconnectedBufferOptions.setBufferSize(100);
                    disconnectedBufferOptions.setPersistBuffer(false);
                    disconnectedBufferOptions.setDeleteOldestMessages(false);
                    mqttAndroidClient.setBufferOpts(disconnectedBufferOptions);
                    subscribeToTopic();
                }

                @Override
                public void onFailure(IMqttToken asyncActionToken, Throwable exception) {
                    mHubToSdkCallback.onConnectionStateChange(false);
                    iotSDKLogUtils.log(true, isDebug, "ERR_IN13", context.getString(R.string.ERR_IN13));
                }
            });

        } catch (MqttException ex) {
            ex.printStackTrace();
        }
    }

    public void subscribeToTopic() {
        iotSDKLogUtils.log(false, this.isDebug, "INFO_IN05", context.getString(R.string.INFO_IN05));
        try {
            String[] topics = {subscriptionTopic, TWIN_SUB_TOPIC, TWIN_SUB_TOPIC_BLANK_MSG};
            mqttAndroidClient.subscribe(topics, new int[]{0, 0, 0}, null, new IMqttActionListener() {
                @Override
                public void onSuccess(IMqttToken asyncActionToken) {
                    mHubToSdkCallback.onConnectionStateChange(true);
                }

                @Override
                public void onFailure(IMqttToken asyncActionToken, Throwable exception) {
                    mHubToSdkCallback.onConnectionStateChange(false);
                }
            });

        } catch (MqttException ex) {
            ex.printStackTrace();
        }
    }

    public void disconnectClient() {
        if (mqttAndroidClient != null) {
            try {
                mqttAndroidClient.disconnect();
            } catch (MqttException e) {
                e.printStackTrace();
            }

            mHubToSdkCallback.onConnectionStateChange(false);
            mqttAndroidClient.unregisterResources();
//            mqttAndroidClient.close();
            mqttAndroidClient = null;
        }
    }

    public void getAllTwins() {
        if (mqttAndroidClient != null && mqttAndroidClient.isConnected()) {
            try {
                MqttMessage message = new MqttMessage();
                message.setPayload("".getBytes());
                mqttAndroidClient.publish(TWIN_PUB_TOPIC_BLANK_MSG, message);


            } catch (MqttException e) {
                e.printStackTrace();
            }
        }
    }

    public void updateTwin(String msgPublish) {

        if (mqttAndroidClient != null && mqttAndroidClient.isConnected()) {
            try {
                MqttMessage message = new MqttMessage();
                message.setPayload(msgPublish.getBytes());
                mqttAndroidClient.publish(TWIN_PUB_TOPIC, message);


                iotSDKLogUtils.log(false, this.isDebug, "INFO_TP01", context.getString(R.string.INFO_TP01) + " " + IotSDKUtils.getCurrentDate());
            } catch (MqttException e) {
                e.printStackTrace();
            }

        }
    }

    public void publishMessage(String msgPublish) {
        try {
            if (mqttAndroidClient != null && mqttAndroidClient.isConnected() && msgPublish != null) {
                MqttMessage message = new MqttMessage();
                message.setPayload(msgPublish.getBytes());
                mqttAndroidClient.publish(publishTopic, message);
                mHubToSdkCallback.onSendMsg(msgPublish);

                iotSDKLogUtils.log(false, this.isDebug, "INFO_SD01", context.getString(R.string.INFO_SD01) + " " + IotSDKUtils.getCurrentDate());
            } else {
                iotSDKLogUtils.log(true, this.isDebug, "ERR_SD10", context.getString(R.string.ERR_SD10) + " : " + IotSDKUtils.getCurrentDate());
            }
        } catch (MqttException e) {
            e.printStackTrace();
        }
    }
}

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/mqtt/IotSDKMQTTService.kt
package com.iotconnectsdk.mqtt

import android.content.Context
import android.os.Build
import android.text.TextUtils
import com.iotconnectsdk.BuildConfig
import com.iotconnectsdk.R
import com.iotconnectsdk.enums.BrokerType
import com.iotconnectsdk.interfaces.HubToSdkCallback
import com.iotconnectsdk.interfaces.PublishMessageCallback
import com.iotconnectsdk.interfaces.TwinUpdateCallback
import com.iotconnectsdk.utils.DateTimeUtils
import com.iotconnectsdk.utils.IotSDKLogUtils
import com.iotconnectsdk.utils.IotSDKUrls
import com.iotconnectsdk.utils.SDKClientUtils.generateSasToken
import com.iotconnectsdk.utils.SDKClientUtils.generateSasTokenLatest
import com.iotconnectsdk.utils.SecurityHelper.createSocketFactory
import com.iotconnectsdk.webservices.responsebean.IdentityServiceResponse
import org.eclipse.paho.android.service.MqttAndroidClient
import org.eclipse.paho.client.mqttv3.*
import org.json.JSONObject


/**
 * Service class for MQTT data upload to cloud
 */
internal class IotSDKMQTTService private constructor(
    private val context: Context,
    private val sdkOptions: String?,
    private val protocolBean: IdentityServiceResponse.D.P,
    private val authenticationType: Int,
    private val hubToSdkCallback: HubToSdkCallback,
    private val publishMessageCallback: PublishMessageCallback,
    private val twinCallbackMessage: TwinUpdateCallback,
    private val iotSDKLogUtils: IotSDKLogUtils,
    private val isDebug: Boolean,
    private val uniqueId: String,
    private val cpId: String,
) {

    private var mqttAndroidClient: MqttAndroidClient? = null

    //- To receive the desired property and update the Reported Property
    private var TWIN_SHADOW_PUB_TOPIC = ""

    private var TWIN_SHADOW_SUB_TOPIC = ""

    // - To Publish the blank message on publish and Subscribe message with desired and reported property
    private var TWIN_SHADOW_PUB_TOPIC_BLANK_MSG = ""

    private var TWIN_SHADOW_SUB_TOPIC_BLANK_MSG = ""

    private val DESIRED = "desired"

    private val UNIQUE_ID = "uniqueId"

    private var subscriptionTopic: String? = null


    private val CPID_DEVICEID = "$cpId-$uniqueId"


    companion object {

        @Volatile
        private var iotSDKMQTTService: IotSDKMQTTService? = null
        fun getInstance(
            context: Context, sdkOptions: String?, protocolBean: IdentityServiceResponse.D.P,
            authenticationType: Int, hubToSdkCallback: HubToSdkCallback,
            publishMessageCallback: PublishMessageCallback, twinCallbackMessage: TwinUpdateCallback,
            iotSDKLogUtils: IotSDKLogUtils, isDebug: Boolean, uniqueId: String, cpId: String
        ): IotSDKMQTTService? {

            synchronized(this) {
                if (iotSDKMQTTService == null) {
                    iotSDKMQTTService = IotSDKMQTTService(
                        context,
                        sdkOptions,
                        protocolBean,
                        authenticationType,
                        hubToSdkCallback,
                        publishMessageCallback,
                        twinCallbackMessage,
                        iotSDKLogUtils,
                        isDebug,
                        uniqueId,
                        cpId
                    )
                }
                return iotSDKMQTTService
            }

        }
    }

    fun clearInstance() {
        iotSDKMQTTService = null
    }


    /*connect MQTT*/
    fun connectMQTT() {
        //init log.
        iotSDKLogUtils.log(false, isDebug, "INFO_IN04", context.getString(R.string.INFO_IN04))
        subscriptionTopic = protocolBean.topics.c2d
        mqttAndroidClient = MqttAndroidClient(
            context, "ssl://" + protocolBean.h + ":" + protocolBean.p, protocolBean.id
        )
        mqttAndroidClient?.setCallback(object : MqttCallbackExtended {
            override fun connectComplete(reconnect: Boolean, serverURI: String) {
                // Because Clean Session is true, we need to re-subscribe
                if (reconnect) {
                    if (subscriptionTopic != null) {
                        subscribeToTopic()
                    }
                }

            }

            override fun connectionLost(cause: Throwable?) {
            }

            @Throws(Exception::class)
            override fun messageArrived(topic: String, message: MqttMessage) {
                try {
                    if (message.payload.toString().isEmpty()) {
                        return
                    }
                    if (topic.contains(
                            TWIN_SHADOW_SUB_TOPIC.substring(
                                0,
                                TWIN_SHADOW_SUB_TOPIC.length - 1
                            )
                        )
                    ) {
                        val mainObj = JSONObject()
                        mainObj.put(DESIRED, JSONObject(String(message.payload)))
                        mainObj.put(UNIQUE_ID, uniqueId)
                        twinCallbackMessage.twinUpdateCallback(mainObj)
                    } else if (topic.contains(
                            TWIN_SHADOW_SUB_TOPIC_BLANK_MSG.substring(
                                0,
                                TWIN_SHADOW_SUB_TOPIC_BLANK_MSG.length - 1
                            )
                        )
                    ) {
                        val mainObj = JSONObject(String(message.payload))
                        mainObj.put(UNIQUE_ID, uniqueId)
                        twinCallbackMessage.twinUpdateCallback(mainObj)
                    } else {
                        hubToSdkCallback.onReceiveMsg(String(message.payload))
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }

            override fun deliveryComplete(token: IMqttDeliveryToken) {
            }
        })
        val mqttConnectOptions = MqttConnectOptions()
        mqttConnectOptions.isAutomaticReconnect = true
        mqttConnectOptions.isCleanSession = true

        var sdkObj: JSONObject
        try {
            if (sdkOptions != null) {
                sdkObj = JSONObject(sdkOptions)

                if (BuildConfig.BrokerType == BrokerType.AZ.value) {
                    TWIN_SHADOW_PUB_TOPIC = "\$iothub/twin/PATCH/properties/reported/?\$rid=1"
                    TWIN_SHADOW_SUB_TOPIC = "\$iothub/twin/PATCH/properties/desired/#"

                    TWIN_SHADOW_PUB_TOPIC_BLANK_MSG = "\$iothub/twin/GET/?\$rid=0"
                    TWIN_SHADOW_SUB_TOPIC_BLANK_MSG = "\$iothub/twin/res/#"
                } else if (BuildConfig.BrokerType == BrokerType.AWS.value) {
                    TWIN_SHADOW_PUB_TOPIC =
                        "\$rid=1\$aws/things/$CPID_DEVICEID/shadow/name/${CPID_DEVICEID}_twin_shadow/report"
                    TWIN_SHADOW_SUB_TOPIC =
                        "\$aws/things/$CPID_DEVICEID/shadow/name/${CPID_DEVICEID}_twin_shadow/property-shadow"

                    TWIN_SHADOW_PUB_TOPIC_BLANK_MSG =
                        "\$aws/things/$CPID_DEVICEID/shadow/name/${CPID_DEVICEID}_twin_shadow/get"
                    TWIN_SHADOW_SUB_TOPIC_BLANK_MSG =
                        "\$aws/things/$CPID_DEVICEID/shadow/name/${CPID_DEVICEID}_twin_shadow/get/all"
                } else {
                    TWIN_SHADOW_PUB_TOPIC = "\$iothub/twin/PATCH/properties/reported/?\$rid=1"
                    TWIN_SHADOW_SUB_TOPIC = "\$iothub/twin/PATCH/properties/desired/#"

                    TWIN_SHADOW_PUB_TOPIC_BLANK_MSG = "\$iothub/twin/GET/?\$rid=0"
                    TWIN_SHADOW_SUB_TOPIC_BLANK_MSG = "\$iothub/twin/res/#"
                }

                if (authenticationType == IotSDKUrls.AUTH_TYPE_SELF_SIGN || authenticationType == IotSDKUrls.AUTH_TYPE_CA_SIGN ||
                    authenticationType == IotSDKUrls.AUTH_TYPE_CA_SIGN_INDI
                ) {
                    if (sdkObj.has("certificate")) {
                        val certificate = sdkObj.getJSONObject("certificate")

                        var caFile: String? = null
                        var clientCrtFile: String? = null
                        var clientKeyFile: String? = null



                        if (certificate.has("SSLCaPath")) {
                            caFile = certificate.getString("SSLCaPath")
                        }

                        if (certificate.has("SSLCertPath")) {
                            clientCrtFile = certificate.getString("SSLCertPath")
                        }

                        if (certificate.has("SSLKeyPath")) {
                            clientKeyFile = certificate.getString("SSLKeyPath")
                        }
                        val clientKeyPassword = ""
                        if (TextUtils.isEmpty(caFile) && TextUtils.isEmpty(clientCrtFile)
                            && TextUtils.isEmpty(clientKeyFile)
                        ) {
                            return
                        }

                        val socketFactory = createSocketFactory(
                            caFile!!,
                            clientCrtFile!!,
                            clientKeyFile!!,
                            "",
                            "",
                            ""
                        )
                        mqttConnectOptions.socketFactory = socketFactory
                    }

                } else if (authenticationType == IotSDKUrls.AUTH_TYPE_SYMMETRIC_KEY) {
                    if (sdkObj.has("devicePK")) {
                        val devicePK = sdkObj.getString("devicePK")
                        if (TextUtils.isEmpty(devicePK)) {
                            return
                        }

                        val resourceUri = "${protocolBean.h}/devices/${protocolBean.id}"
                        val generateSasToken = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                            generateSasTokenLatest(resourceUri, devicePK)
                        } else {
                            generateSasToken(resourceUri, devicePK)
                        }
                        if (generateSasToken != null) {
                            mqttConnectOptions.password = generateSasToken.toCharArray()
                        }
                    }
                } else if (authenticationType == IotSDKUrls.AUTH_TYPE_TOKEN) {
                    if (protocolBean.pwd != null) {
                        mqttConnectOptions.password = protocolBean.pwd.toCharArray()
                    }
                }

            }

        } catch (e: Exception) {
            e.printStackTrace()
        }

        mqttConnectOptions.userName = protocolBean.un

        try {
            mqttAndroidClient?.connect(mqttConnectOptions, null, object : IMqttActionListener {
                override fun onSuccess(asyncActionToken: IMqttToken) {
                    val disconnectedBufferOptions = DisconnectedBufferOptions()
                    disconnectedBufferOptions.isBufferEnabled = true
                    disconnectedBufferOptions.bufferSize = 100
                    disconnectedBufferOptions.isPersistBuffer = false
                    disconnectedBufferOptions.isDeleteOldestMessages = false
                    mqttAndroidClient?.setBufferOpts(disconnectedBufferOptions)
                    subscribeToTopic()

                }

                override fun onFailure(asyncActionToken: IMqttToken, exception: Throwable) {
                    hubToSdkCallback.onConnectionStateChange(false)
                    iotSDKLogUtils.log(
                        true, isDebug, "ERR_IN13", context.getString(R.string.ERR_IN13)
                    )
                }
            })
        } catch (ex: MqttException) {
            ex.printStackTrace()
        }
    }

    /*Subscribe to specific topic*/
    fun subscribeToTopic() {
        iotSDKLogUtils.log(false, isDebug, "INFO_IN05", context.getString(R.string.INFO_IN05))
        try {
            val topics =
                arrayOf(subscriptionTopic, TWIN_SHADOW_SUB_TOPIC, TWIN_SHADOW_SUB_TOPIC_BLANK_MSG)
            mqttAndroidClient?.subscribe(topics, intArrayOf(0, 0, 0), null,
                object : IMqttActionListener {
                    override fun onSuccess(asyncActionToken: IMqttToken) {
                        hubToSdkCallback.onConnectionStateChange(true)
                        publishMessageCallback.onSendMsg()
                    }

                    override fun onFailure(asyncActionToken: IMqttToken, exception: Throwable) {
                        hubToSdkCallback.onConnectionStateChange(false)
                    }
                })
        } catch (ex: MqttException) {
            ex.printStackTrace()
        }
    }

    /*Disconnect from MQTT client*/
    fun disconnectClient() {
        if (mqttAndroidClient != null) {
            try {
                mqttAndroidClient?.disconnect()
            } catch (e: MqttException) {
                e.printStackTrace()
            }
            mqttAndroidClient?.unregisterResources()
            mqttAndroidClient = null
            hubToSdkCallback.onConnectionStateChange(false)
        }
    }

    /*get all twins from portal*/
    fun getAllTwins() {
        if (mqttAndroidClient != null && mqttAndroidClient!!.isConnected) {
            try {
                val message = MqttMessage()
                message.payload = "".toByteArray()
                mqttAndroidClient!!.publish(TWIN_SHADOW_PUB_TOPIC_BLANK_MSG, message)
            } catch (e: MqttException) {
                e.printStackTrace()
            }
        }
    }

    /*update twin in there is any changes*/
    fun updateTwin(msgPublish: String) {
        if (mqttAndroidClient != null && mqttAndroidClient!!.isConnected) {
            try {
                val message = MqttMessage()
                message.payload = msgPublish.toByteArray()
                mqttAndroidClient!!.publish(TWIN_SHADOW_PUB_TOPIC, message)
                iotSDKLogUtils.log(
                    false, isDebug, "INFO_TP01",
                    context.getString(R.string.INFO_TP01) + " " + DateTimeUtils.currentDate
                )
            } catch (e: MqttException) {
                e.printStackTrace()
            }
        }
    }


    /*publish message on portal on different topics*/
    fun publishMessage(topics: String, msgPublish: String?) {
        try {
            if (mqttAndroidClient != null && mqttAndroidClient!!.isConnected && msgPublish != null) {
                val message = MqttMessage()
                message.payload = msgPublish.toByteArray()
                mqttAndroidClient?.publish(topics, message)
                hubToSdkCallback.onSendMsgUI(msgPublish)
                iotSDKLogUtils.log(
                    false, isDebug, "INFO_SD01",
                    context.getString(R.string.INFO_SD01) + " " + DateTimeUtils.currentDate
                )
            } else {
                iotSDKLogUtils.log(
                    true, isDebug, "ERR_SD10",
                    context.getString(R.string.ERR_SD10) + " : " + DateTimeUtils.currentDate
                )
            }
        } catch (e: MqttException) {
            e.printStackTrace()
        }
    }

}